ts/tat-trend.ts

```diff
@@ -0,0 +1,437 @@
+  const month = date.getMonth() + 1;
+  let fiscalYear = year;
+  let fiscalQuarter: number;
+
+  if (month >= 4) {
+    fiscalYear = year + 1; // next fiscal year
+    fiscalQuarter = Math.floor((month - 4) / 3) + 1;
+  } else {
+    fiscalYear = year; // current fiscal year
+    fiscalQuarter = Math.floor((month + 12 - 4) / 3) + 1;
+  }
+  const fiscalYearString = `FY${fiscalYear - 1}/${String(fiscalYear).slice(
+    -2
+  )}`;
+  switch (selectedGrouping) {
+    case "week":
+      return getWeek(date);
+    case "month":
+      return `${fiscalYearString} - ${String(month).padStart(2, "0")}`;
+    case "fiscalQuarter":
+      return `${fiscalYearString} Q${fiscalQuarter}`;
+    case "fiscalYear":
+      return fiscalYearString;
+    default:
+      throw new Error(`Unsupported grouping type: ${selectedGrouping}`);
+  }
+}
+
+function groupData(
+  jsonData: any[],
+  selectedGrouping: string,
+  selectedGates: string[],
+  selectedDataType: string,
+  includeIncomplete: boolean = false
+): AssayGroups {
+  const assayGroups: AssayGroups = {};
+
+  jsonData.forEach((row: any) => {
+    const assay = row["Assay"];
+    const caseId = row["Case ID"];
+    const caseCompletedDate =
+      row[`${selectedDataType} Release Completed`] ?? row[GATE.ALL_COMPLETED];
+    const totalDays =
+      row[`${selectedDataType} Total Days`] ?? row["ALL Total Days"];
+
+    // skip if case completion date is missing and includeIncomplete is false
+    if (!caseCompletedDate && !includeIncomplete) {
+      return;
+    }
+    if (assay && totalDays != null && caseId != null) {
+      if (!assayGroups[assay]) {
+        assayGroups[assay] = {};
+      }
+      selectedGates.forEach((gate) => {
+        const { completedDate, days } = getCompletedDateAndDays(
+          row,
+          gate,
+          selectedDataType
+        );
+        if (!completedDate) {
+          return;
+        }
+        const date = new Date(caseCompletedDate || completedDate);
+        const group = getGroup(date, selectedGrouping);
+        if (!assayGroups[assay][gate]) {
+          assayGroups[assay][gate] = { x: [], y: [], text: [], n: 0 };
+        }
+        assayGroups[assay][gate].x.push(group);
+        assayGroups[assay][gate].y.push(days ?? totalDays);
+        assayGroups[assay][gate].text.push(
+          `${caseId} (${gate === "All Completed" ? "All Completed" : gate})`
+        );
+        assayGroups[assay][gate].n += 1;
+      });
+    }
+  });
+  return assayGroups;
+}
+
+function getWeek(date: Date): string {
+  const onejan = new Date(date.getFullYear(), 0, 1);
+  const week = Math.ceil(
+    ((date.getTime() - onejan.getTime()) / 86400000 + onejan.getDay() + 1) / 7
+  );
+  return `${date.getFullYear()}-W${String(week).padStart(2, "0")}`;
+}
+
+function getColorByGate(): boolean {
+  const toggleColors = document.getElementById(
+    "toggleColors"
+  ) as HTMLInputElement;
+  return toggleColors ? toggleColors.checked : false;
```

Comment:
`toggleColors` should never be undefined here, right? Better to throw an error rather than handle an unexpected state in an unexpected way